Remove error code types

majewsky opened the pull request:
This is a breaking change intended for the v2.0 release, spun off from the discussion in #2902. I am posting this as a bit of a hot take. If this ends up being a discussion about why the ErrDefaultXXX types and the ErrorContext field are actually useful, and we just end up with actual real documentation on these, that would be a good outcome as well.

## Overview and rationale

The ErrDefault series of types (ErrDefault400, ErrDefault401, etc.) serve the apparent purpose of allowing to assert on a specific response code (e.g. to handle 404 responses differently from other 4xx/5xx responses). However, because of their design, they often tend to obscure the real underlying error message in many cases, reporting just something like "Method not allowed" instead of the default error string on ErrUnexpectedResponseCode that explains which request failed and includes the actual error message returned by the OpenStack service.

This commit removes the types, and all of the associated boilerplate, and replaces them with a single small helper function for quickly asserting that an error is an ErrUnexpectedResponseCode with a specific response code. A similar design has been in productive use in [my Swift client library](https://github.com/majewsky/schwift/blob/v1.3.0/errors.go#L152-L157) for several years.

This also gets rid of the undocumented and untested ProviderClient.ErrorContext interface that I have legitimately no idea what it's supposed to be good for, besides maybe allowing users to skip the generation of ErrDefaultXXX-type errors. If that's the only purpose, that would be yet more incentive to just get rid of them.

Finally, as a small (but still breaking) change, this removes some newlines from the error messages generated by ErrUnexpectedResponseCode. Newlines in error messages cause several log aggregators to choke (e.g. I have seen problems with this myself in syslog and fluentd), so we now try a bit harder to avoid them if possible.

pierreprinetti commented:
I am working on leveraging errors.Unwrap, which should bring the same benefits: https://github.com/gophercloud/gophercloud/pull/2906

pierreprinetti commented:
I completely agree that (RequestOpts).ContextError must go, since error unwrapping should provide the same capability in a more idiomatic way.

I am finding myself in deeper a rabbit hole than I can handle with #2906. If this change fits your needs, then I'd be inclined to merge.

pierreprinetti commented:
@gophercloud/core what do you think?

mandre reviewed (approved):
I'm very much in favor of this change as it makes consuming gophercloud easier.

pierreprinetti commented:
@majewsky Can you please rebase? I think we can then merge :rocket:

majewsky commented:
rebase done

pierreprinetti commented:
hmm what happened to the CI

pierreprinetti commented:
@majewsky something wrong happened on the Github side. When checking out this branch locally, I get the old commits and not the ones you just force-pushed. Do you mind `git commit --amend -no-edit && git push --force` please?

majewsky commented:
I rebased as requested three weeks ago, but apparently this was not seen (maybe because I did not also comment "done"?). Because new merge conflicts have arisen, I have rebased again.

majewsky commented:
Rebased onto current master to match 2.0.0-beta.4. Please take another look.

mandre commented:
There's a failed conflict resolution in `internal/acceptance/openstack/identity/v3/federation_test.go`.

majewsky commented:
One error remains in a testcase that I did not touch (`TestBGPAgentRUD`). The concrete failure does not seem to relate to my changes, so I'll have to leave it to you to decide if this is caused by the PR.
